### PR TITLE
Add ability to use a package.json for node applications

### PR DESCRIPTION
**Pull request by dstgermain**
This enables the use of package.json files to define attributes.
Removed Pom file functionality and Package file from check_params
to decrease the cyclomatic complexity.

### CODE REVIEW

**Comment by athieriot** [comment]
Hi @dstgermain,

   Thank you so much for your work, I was really curious to know how other kind of binary would work with that plugin.

   I am going to add a few comments that are mostly aesthetic but it looks great (Nice tests !)

   One other thing that would be nice for you to do if you don't mind, is to add a few lines about that feature in the ```DOCS.md``` file. So it can be picked up in the website.

File: index.js [hunk 1]
```diff
@@ -61,6 +61,57 @@ var do_upload = function (params) {
   );
 }
 
+var parse_pom_file = function (params, resolve, reject) {
```
Comment (athieriot) [inline index.js]: Basically, I would simply returns a Promise there and not pass resolve nor reject as argument.

We use the same pattern in different part of the code:

```
return new Promise((resolve, reject) => {
}
```

It is as well going to make testing simpler

File: index.js [hunk 2]
```diff
@@ -61,6 +61,57 @@ var do_upload = function (params) {
...
+
+    if(params.vargs.files.indexOf(params.vargs.pom)==-1) {
+      params.vargs.files.push(params.vargs.pom);
+    }
+
+    return resolve(params);
+  });
+}
+
+var parse_package_file = function (params, resolve, reject) {
```
Comment (athieriot) [inline index.js]: Same here

File: index.js [hunk 3]
```diff
@@ -61,6 +61,57 @@ var do_upload = function (params) {
...
+  }
+
+  pomParser.parse({ filePath: params.workspace.path + '/' + params.vargs.pom }, function(err, pomResponse) {
+    if (err) { return reject('An error happened while trying to parse the pom file: ' + err); }
+
+    params.vargs.group_id    || (params.vargs.group_id = pomResponse.pomObject.project.groupid);
+    params.vargs.artifact_id || (params.vargs.artifact_id = pomResponse.pomObject.project.artifactid);
+    params.vargs.version     || (params.vargs.version = pomResponse.pomObject.project.version);
+
+    if (!params.vargs.group_id || !params.vargs.artifact_id || !params.vargs.version) {
```
Comment (athieriot) [inline index.js]: This is not major but would you mind looking at a way to share those ~7 lines of code between package and pom methods?

Comment (dstgermain) [inline index.js]: makes sense, I'll work on that.

File: index.js (hunk 1 above)
Comment (dstgermain) [inline index.js]: Sure no problem.

**Comment by dstgermain** [comment]
I was about to add to the md when I got distracted. I'll Add those changes and update the md. Do you prefer squashed commits?

**Comment by athieriot** [comment]
That would be great if you can squash your commits yeah.

The change is pretty concise so it's probably clearer that way :)

File: index.js [hunk 4]
```diff
@@ -61,6 +61,67 @@ var do_upload = function (params) {
...
+    if (!fs.existsSync(params.workspace.path + '/' + params.vargs.pom)) {
+      return reject('Given pom file has to exists');
+    }
+
+    pomParser.parse({ filePath: params.workspace.path + '/' + params.vargs.pom }, function(err, pomResponse) {
+      if (err) { return reject('An error happened while trying to parse the pom file: ' + err); }
+      params.vargs.files || (params.vargs.files = []);
+      params.vargs = merge_arguments(params.vargs, pomResponse.pomObject.project);
+
+      if (!params.vargs.group_id || !params.vargs.artifact_id || !params.vargs.version) {
```
Comment (athieriot) [inline index.js]: I was actually talking about reusing those ~7 lines from 82 to 88 :)
What do you think?

Comment (dstgermain) [inline index.js]: Oh I see, yup that makes sense.

**Comment by athieriot** [comment]
Ola !
Just wanted to know how you are going along with that.

You are so close it would be a shame not to merge your work !

**Pull request closed without merging**